utils/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/jsonpayload/JsonPayloadToAvroConverter.scala

```diff
@@ -0,0 +1,146 @@
+
+  object DateConverter extends AvroTypeConverter {
+    private val conversion = new TimeConversions.DateConversion
+
+    override def canManage(schema: Schema, path: util.Deque[String]): Boolean =
+      schema.getType == Schema.Type.INT && schema.getLogicalType == LogicalTypes.date()
+
+    override def convert(field: Schema.Field, schema: Schema, jsonValue: AnyRef, path: util.Deque[String], silently: Boolean): AnyRef = {
+      jsonValue match {
+        case number: Number => conversion.fromInt(number.intValue(), schema, schema.getLogicalType)
+        case str: String => Decoder[LocalDate].decodeJson(fromString(str)).fold(throw _, identity)
+        case other => other
+      }
+    }
+  }
+
+  object TimeMillisConverter extends AvroTypeConverter {
+    private val conversion = new TimeConversions.TimeMillisConversion
+
+    override def canManage(schema: Schema, path: util.Deque[String]): Boolean =
+      schema.getType == Schema.Type.INT && schema.getLogicalType == LogicalTypes.timeMillis()
+
+    override def convert(field: Schema.Field, schema: Schema, jsonValue: AnyRef, path: util.Deque[String], silently: Boolean): AnyRef = {
+      jsonValue match {
+        case number: Number => conversion.fromInt(number.intValue(), schema, schema.getLogicalType)
+        case str: String => Decoder[LocalTime].decodeJson(fromString(str)).fold(throw _, identity)
+        case other => other
+      }
+    }
+  }
+
+  object TimeMicrosConverter extends AvroTypeConverter {
+    private val conversion = new TimeConversions.TimeMicrosConversion
+
+    override def canManage(schema: Schema, path: util.Deque[String]): Boolean =
+      schema.getType == Schema.Type.LONG && schema.getLogicalType == LogicalTypes.timeMicros()
+
+    override def convert(field: Schema.Field, schema: Schema, jsonValue: AnyRef, path: util.Deque[String], silently: Boolean): AnyRef = {
+      jsonValue match {
+        case number: Number => conversion.fromLong(number.intValue(), schema, schema.getLogicalType)
+        case str: String => Decoder[LocalTime].decodeJson(fromString(str)).fold(throw _, identity)
+        case other => other
+      }
+    }
+  }
+
+  object TimestampMillisConverter extends AvroTypeConverter {
+    private val conversion = new TimeConversions.TimestampMillisConversion
+
+    override def canManage(schema: Schema, path: util.Deque[String]): Boolean =
+      schema.getType == Schema.Type.LONG && schema.getLogicalType == LogicalTypes.timestampMillis()
+
+    override def convert(field: Schema.Field, schema: Schema, jsonValue: AnyRef, path: util.Deque[String], silently: Boolean): AnyRef = {
+      jsonValue match {
+        case number: Number => conversion.fromLong(number.intValue(), schema, schema.getLogicalType)
+        case str: String => Decoder[Instant].decodeJson(fromString(str)).fold(throw _, identity)
+        case other => other
+      }
+    }
+  }
+
+  object TimestampMicrosConverter extends AvroTypeConverter {
```

Comment:
all those methods look quite similar (in terms of convert method)? Shouldn't we extract sth common?

Reply:
I've extracted something - we can't do too much with this